dialogue-core/src/test/java/com/palantir/dialogue/core/NodeSelectionStrategyChannelTest.java

```diff
@@ -0,0 +1,107 @@
+import static org.mockito.ArgumentMatchers.eq;
+import static org.mockito.Mockito.lenient;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.times;
+import static org.mockito.Mockito.verify;
+
+import com.github.benmanes.caffeine.cache.Ticker;
+import com.google.common.collect.ImmutableList;
+import com.google.common.util.concurrent.Futures;
+import com.palantir.dialogue.Response;
+import com.palantir.tritium.metrics.registry.DefaultTaggedMetricRegistry;
+import java.util.List;
+import java.util.Optional;
+import java.util.Random;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.ExtendWith;
+import org.mockito.Mock;
+import org.mockito.Mockito;
+import org.mockito.Spy;
+import org.mockito.junit.jupiter.MockitoExtension;
+
+@ExtendWith(MockitoExtension.class)
+class NodeSelectionStrategyChannelTest {
+
+    @Spy
+    private NodeSelectionStrategyChooser strategySelector = new NodeSelectionStrategyChooser() {
+        @Override
+        public Optional<DialogueNodeSelectionStrategy> updateAndGet(
+                List<DialogueNodeSelectionStrategy> updatedStrategies) {
+            return NodeSelectionStrategyChannel.getFirstKnownStrategy(updatedStrategies);
+        }
+    };
+
+    @Mock
+    private LimitedChannel channel1;
+
+    @Mock
+    private LimitedChannel channel2;
+
+    @Mock
+    private Ticker clock;
+
+    private String channelName = "channelName";
+    private Random pseudo = new Random(12893712L);
+    private NodeSelectionStrategyChannel channel;
+
+    @BeforeEach
+    void beforeEach() {
+        channel = new NodeSelectionStrategyChannel(
+                strategySelector,
+                DialogueNodeSelectionStrategy.PIN_UNTIL_ERROR_WITHOUT_RESHUFFLE,
+                channelName,
+                pseudo,
+                clock,
+                new DefaultTaggedMetricRegistry());
+    }
+
+    @Test
+    void handles_zero_to_one_uri_update() {
+        channel.updateChannels(ImmutableList.of());
+        channel.updateChannels(ImmutableList.of(channel1));
+    }
+
+    @Test
+    void handles_one_to_many_uri_update() {
+        channel.updateChannels(ImmutableList.of(channel1));
+        channel.updateChannels(ImmutableList.of(channel1, channel2));
+    }
+
+    @Test
+    void updates_strategy_on_response() {
+        channel.updateChannels(ImmutableList.of(channel1));
+        setResponse(channel1, Optional.of("BALANCED,FOO"));
+        channel.maybeExecute(null, null).get();
+        verify(strategySelector, times(1))
+                .updateAndGet(eq(ImmutableList.of(
+                        DialogueNodeSelectionStrategy.BALANCED, DialogueNodeSelectionStrategy.UNKNOWN)));
+    }
+
+    private static void setResponse(LimitedChannel mockChannel, Optional<String> header) {
+        Mockito.clearInvocations(mockChannel);
+        Mockito.reset(mockChannel);
+        Response resp = mock(Response.class);
```

Comment:
would prefer to use TestResponse.withheader rather than mocking, but not a big deal